no method named `capture` found for struct `Vec<Screen>` in the current scope

error[E0599]: no method named `capture` found for struct `Vec<Screen>` in the current scope
  --> src/main.rs:10:28
   |
10 |     let mut image = screen.capture().unwrap();
   |                            ^^^^^^^ method not found in `Vec<Screen>`

error[E0609]: no field `id` on type `Vec<Screen>`
  --> src/main.rs:12:47
   |
12 |     fs::write(format!("target/{}.png", screen.id.to_string()), &buffer).unwrap();
   |                                               ^^ unknown field

error[E0599]: no method named `capture_area` found for struct `Vec<Screen>` in the current scope
  --> src/main.rs:14:20
   |
14 |     image = screen.capture_area(300, 300, 300, 300).unwrap();
   |                    ^^^^^^^^^^^^ method not found in `Vec<Screen>`

error[E0609]: no field `id` on type `Vec<Screen>`
  --> src/main.rs:16:49
   |
16 |     fs::write(format!("target/{}-2.png", screen.id.to_string()), &buffer).unwrap();
   |                                                 ^^ unknown field

Some errors have detailed explanations: E0599, E0609.
For more information about an error, try `rustc --explain E0599`.
error: could not compile `screen_shot` due to 4 previous errors

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "screenshots"
-version = "0.4.4"
+version = "0.4.5"
 edition = "2021"
 description = "A cross-platform screen capturer library"
 license = "Apache-2.0"

--- a/README.md
+++ b/README.md
@@ -10,17 +10,17 @@ use std::{fs, time::Instant};
 
 fn main() {
   let start = Instant::now();
-  let screens = Screen::all();
+  let screens = Screen::all().unwrap();
 
-  for screen in screens.unwrap() {
+  for screen in screens {
     println!("capturer {:?}", screen);
     let mut image = screen.capture().unwrap();
     let mut buffer = image.buffer();
-    fs::write(format!("target/{}.png", screen.display_info.id.to_string()), &buffer).unwrap();
+    fs::write(format!("target/{}.png", screen.display_info.id), &buffer).unwrap();
 
     image = screen.capture_area(300, 300, 300, 300).unwrap();
     buffer = image.buffer();
-    fs::write(format!("target/{}-2.png", screen.display_info.id.to_string()), &buffer).unwrap();
+    fs::write(format!("target/{}-2.png", screen.display_info.id), &buffer).unwrap();
   }
 
   let screen = Screen::from_point(100, 100).unwrap();

--- a/src/darwin.rs
+++ b/src/darwin.rs
@@ -1,5 +1,5 @@
 use crate::{DisplayInfo, Image};
-use core_graphics::display::{CGDisplay, CGPoint, CGRect, CGSize};
+use core_graphics::display::CGDisplay;
 
 pub fn capture_screen(display_info: &DisplayInfo) -> Option<Image> {
   let cg_display = CGDisplay::new(display_info.id);
@@ -21,22 +21,26 @@ pub fn capture_screen_area(
   height: u32,
 ) -> Option<Image> {
   let cg_display = CGDisplay::new(display_info.id);
-  let full_cg_image = cg_display.image()?;
+  let cg_image = cg_display.image()?;
 
-  let w = width as f32 * display_info.scale_factor;
-  let h = height as f32 * display_info.scale_factor;
+  let w = (width as f32 * display_info.scale_factor) as i32;
+  let h = (height as f32 * display_info.scale_factor) as i32;
 
-  let cg_rect = CGRect::new(
-    &CGPoint::new(x as f64, y as f64),
-    &CGSize::new(w as f64, h as f64),
-  );
+  let mut bgra = vec![0; (w * h * 4) as usize];
+  let data = cg_image.data();
+  let bytes = data.bytes();
 
-  let cg_image = full_cg_image.cropped(cg_rect)?;
+  // 图片裁剪
+  for r in y..(y + h) {
+    for c in x..(x + w) {
+      let index = (((r - y) * w + (c - x)) * 4) as usize;
+      let i = ((r * cg_image.width() as i32 + c) * 4) as usize;
+      bgra[index] = bytes[i];
+      bgra[index + 1] = bytes[i + 1];
+      bgra[index + 2] = bytes[i + 2];
+      bgra[index + 3] = bytes[i + 3];
+    }
+  }
 
-  Image::from_bgra(
-    cg_image.width() as u32,
-    cg_image.height() as u32,
-    Vec::from(cg_image.data().bytes()),
-  )
-  .ok()
+  Image::from_bgra(w as u32, h as u32, bgra).ok()
 }

